gnc/pmc/src/pmc_sim.cc

```diff
@@ -0,0 +1,231 @@
+                        Eigen::Matrix<float, 6, 1> & nozzle_area) {
+  const float bpm_servo_max_theta = PMCConstants::abp_nozzle_gear_ratio *
+                                    (PMCConstants::abp_nozzle_max_open_angle - PMCConstants::abp_nozzle_min_open_angle);
+  //  [-] Conversion between servo PWM command (0-100) and resulting angle (0-90)
+  const float bpm_servo_pwm2angle = bpm_servo_max_theta / 255.0;
+  // backlash box
+  for (int i = 0; i < 6; i++) {
+    if (servo_thetas_[i] < backlash_theta_[i] - PMCConstants::backlash_half_width)
+      backlash_theta_[i] = servo_thetas_[i] + PMCConstants::backlash_half_width;
+    else if (servo_thetas_[i] > backlash_theta_[i] + PMCConstants::backlash_half_width)
+      backlash_theta_[i] = servo_thetas_[i] - PMCConstants::backlash_half_width;
+  }
+  Eigen::Matrix<float, 6, 1> voltage,
+    err = bpm_servo_pwm2angle * servo_cmd - backlash_theta_ * PMCConstants::servo_motor_gear_ratio;
+  for (int i = 0; i < 6; i++)
+    voltage[i] = servo_pids_[i].Run(err[i]);
+  servo_current_ = (voltage - PMCConstants::servo_motor_k * prev_omega_) * (1.0 / PMCConstants::servo_motor_r);
+  servo_thetas_ += 1.0 / 62.5 * prev_omega_;
+  servo_thetas_ = servo_thetas_.cwiseMin(bpm_servo_max_theta / PMCConstants::servo_motor_gear_ratio).cwiseMax(0.0);
+  // torque - viscous friction
+  Eigen::Matrix<float, 6, 1> motor_torque =
+    PMCConstants::servo_motor_k * servo_current_ - PMCConstants::servo_motor_friction * prev_omega_;
+  prev_omega_ += (1.0 / PMCConstants::servo_motor_gearbox_inertia) / 62.5 * motor_torque;
+  nozzle_theta = backlash_theta_ * PMCConstants::servo_motor_gear_ratio * 1.0 / PMCConstants::abp_nozzle_gear_ratio;
+  for (int i = 0; i < 6; i++) {
+    //  * nozzle_flap_count * nozzle_width
+    nozzle_area[i] =
+      (PMCConstants::abp_nozzle_intake_height -
+       cos(PMCConstants::abp_nozzle_min_open_angle + nozzle_theta[i]) * PMCConstants::abp_nozzle_flap_length) *
+      2 * PMCConstants::nozzle_widths[i];
+  }
+}
+
+void Blower::ImpellerModel(int speed_cmd, float voltage, float* motor_current, float* motor_torque) {
+  const float max_change = 200*2*M_PI/60.0 * PMCConstants::DT;
+  float speed = speed_cmd / PMCConstants::impeller_speed2pwm;
+  if (speed > prev_speed_rate_ + max_change)
+    speed = prev_speed_rate_ + max_change;
+  else if (speed < prev_speed_rate_ - max_change)
+    speed = prev_speed_rate_ - max_change;
+  prev_speed_rate_ = speed;
+
+  speed = speed - impeller_speed_;
+  float v = speed_pid_.Run(speed);
+  v = std::max(-voltage, std::min(voltage, v));
+
+  *motor_current = (v - impeller_speed_ / PMCConstants::imp_motor_speed_k) / PMCConstants::imp_motor_r;
+  *motor_torque =
+    (*motor_current) * PMCConstants::imp_motor_torque_k - PMCConstants::imp_motor_friction_coeff * impeller_speed_;
+
+  last_speed_ = impeller_speed_;  // matching simulink, but I don't think this is right
```

Comment:
Thinking about this one more I think it's right, it returns the current speed rather than the next speed. I'll remove the comment and create an issue for the other things